src/target/riscv/riscv.c

```diff
@@ -655,20 +664,41 @@ static int set_trigger(struct target *target, unsigned int idx, riscv_reg_t tdat
 		return ERROR_FAIL;
 	if (riscv_reg_get(target, &tdata2_rb, GDB_REGNO_TDATA2) != ERROR_OK)
 		return ERROR_FAIL;
-	bool tdata1_config_denied = (tdata1 & ~tdata1_ignore_mask) != (tdata1_rb & ~tdata1_ignore_mask);
+	bool unsupported_napot_range = false;
+	bool tdata1_config_denied = true;
+	const uint32_t type = get_field(tdata1, CSR_TDATA1_TYPE(riscv_xlen(target)));
+	const bool is_mcontrol = type == CSR_TDATA1_TYPE_MCONTROL;
+	const bool is_napot_match = get_field(tdata1, CSR_MCONTROL_MATCH) == CSR_MCONTROL_MATCH_NAPOT;
+	if (is_mcontrol && is_napot_match) {
+		maskmax = get_field(tdata1_rb, CSR_MCONTROL_MASKMAX(riscv_xlen(target)));
+		const unsigned int napot_size = count_trailing_ones(tdata2) + 1;
+		if (maskmax < napot_size)
+			unsupported_napot_range = true;
+		tdata1_config_denied = (tdata1 & ~maskmax) != (tdata1_rb & ~maskmax);
+	} else if (is_mcontrol && !is_napot_match) {
+		maskmax = get_field(tdata1_rb, CSR_MCONTROL_MASKMAX(riscv_xlen(target)));
+		tdata1_config_denied = (tdata1 & ~maskmax) != (tdata1_rb & ~maskmax);
+	} else {
+		tdata1_config_denied = tdata1 != tdata1_rb;
+	}
 	bool tdata2_config_denied = tdata2 != tdata2_rb;
-	if (tdata1_config_denied || tdata2_config_denied) {
+	if (tdata1_config_denied || tdata2_config_denied || unsupported_napot_range) {
 		LOG_TARGET_DEBUG(target, "Trigger %u doesn't support what we need.", idx);
 
 		if (tdata1_config_denied)
 			LOG_TARGET_DEBUG(target,
-				"After writing 0x%" PRIx64 " to tdata1 it contains 0x%" PRIx64 "; tdata1_ignore_mask=0x%" PRIx64,
-				tdata1, tdata1_rb, tdata1_ignore_mask);
+				"After writing 0x%" PRIx64 " to tdata1 it contains 0x%" PRIx64,
+				tdata1, tdata1_rb);
 
 		if (tdata2_config_denied)
 			LOG_TARGET_DEBUG(target,
-				"wrote 0x%" PRIx64 " to tdata2 but read back 0x%" PRIx64,
+				"After writing 0x%" PRIx64 " to tdata2 it contains 0x%" PRIx64,
 				tdata2, tdata2_rb);
+
+		if (unsupported_napot_range)
+			LOG_TARGET_DEBUG(target,
+				"The requested NAPOT match range (tdata2=0x%" PRIx64 ") exceeds maskmax=0x%" PRIx64,
+				tdata2, maskmax);
 		if (riscv_reg_set(target, GDB_REGNO_TDATA1, 0) != ERROR_OK)
```

Comment:
Detail: Blank line can be used here to maintain consistent style. 

```suggestion
				tdata2, maskmax);

		if (riscv_reg_set(target, GDB_REGNO_TDATA1, 0) != ERROR_OK)
```

Reply:
addressed